File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/backend/connected/aas/submodelelement/ConnectedSubmodelElementCollection.java
package org.eclipse.basyx.aas.backend.connected.aas.submodelelement;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.eclipse.basyx.aas.api.metamodel.aas.qualifier.qualifiable.IConstraint;
import org.eclipse.basyx.aas.api.metamodel.aas.reference.IReference;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.IDataElement;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.ISubmodelElement;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.ISubmodelElementCollection;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.operation.IOperation;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.IProperty;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedVABElementContainerFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasDataSpecificationFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasKindFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasSemanticsFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.QualifiableFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.ReferableFacade;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.IVABElementContainer;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.SubModel;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.SubmodelElementCollection;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.property.Property;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

/**
 * "Connected" implementation of SubmodelElementCollection
 * 
 * @author rajashek
 *
 */
public class ConnectedSubmodelElementCollection extends ConnectedSubmodelElement implements IVABElementContainer, ISubmodelElementCollection {
	ConnectedVABElementContainerFacade facade;

	public ConnectedSubmodelElementCollection(VABElementProxy proxy) {
		super(proxy);
		facade = new ConnectedVABElementContainerFacade(proxy);
	}

	@Override
	public HashSet<IReference> getDataSpecificationReferences() {
		return new HasDataSpecificationFacade(getElem()).getDataSpecificationReferences();
	}

	@Override
	public String getIdshort() {
		return new ReferableFacade(getElem()).getIdshort();
	}

	@Override
	public String getCategory() {
		return new ReferableFacade(getElem()).getCategory();
	}

	@Override
	public String getDescription() {
		return new ReferableFacade(getElem()).getDescription();
	}

	@Override
	public IReference getParent() {
		return new ReferableFacade(getElem()).getParent();
	}

	@Override
	public Set<IConstraint> getQualifier() {
		return new QualifiableFacade(getElem()).getQualifier();
	}

	@Override
	public IReference getSemanticId() {
		return new HasSemanticsFacade(getElem()).getSemanticId();
	}

	@Override
	public String getHasKindReference() {
		return new HasKindFacade(getElem()).getHasKindReference();
	}

	@Override
	public void setValue(ArrayList<?> value) {
		getProxy().setModelPropertyValue(Property.VALUE, value);

	}

	@Override
	public ArrayList<?> getValue() {
		return (ArrayList<?>) getProxy().getModelPropertyValue(Property.VALUE);
	}

	@Override
	public void setOrdered(boolean value) {
		getProxy().setModelPropertyValue(SubmodelElementCollection.ORDERED, value);
	}

	@Override
	public boolean isOrdered() {
		return (boolean) getElem().getPath(SubmodelElementCollection.ORDERED);
	}

	@Override
	public void setAllowDuplicates(boolean value) {
		getProxy().setModelPropertyValue(SubmodelElementCollection.ALLOWDUPLICATES, value);

	}

	@Override
	public boolean isAllowDuplicates() {
		return (boolean) getElem().getPath(SubmodelElementCollection.ALLOWDUPLICATES);
	}

	@Override
	public void setElements(HashMap<String, ISubmodelElement> value) {
		getProxy().setModelPropertyValue(SubModel.SUBMODELELEMENT, value);
	}

	@SuppressWarnings("unchecked")
	@Override
	public HashMap<String, ISubmodelElement> getElements() {
		return (HashMap<String, ISubmodelElement>) getProxy().getModelPropertyValue(SubModel.SUBMODELELEMENT);
	}

	public void addProperty(IProperty property) {
		getElements().put(property.getId(), property);
		getDataElements().put(property.getId(), property);
	}

	public void addOperation(IOperation operation) {
		getElements().put(operation.getId(), operation);
		getOperations().put(operation.getId(), operation);
	}

	@Override
	public void addSubModelElement(ISubmodelElement element) {
		facade.addSubModelElement(element);
	}

	@Override
	public Map<String, IDataElement> getDataElements() {
		return facade.getDataElements();
	}

	@Override
	public Map<String, IOperation> getOperations() {
		return facade.getOperations();
	}
}

File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/ConnectedSubmodelElementCollection.cpp
/*
 * ConnectedSubmodelElementCollection.cpp
 *
 *      Author: wendel
 */

#include "ConnectedSubmodelElementCollection.h"
#include "aas/submodelelement/property/IProperty.h"
#include "aas/ISubModel.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

using namespace basyx::aas::submodelelement;

ConnectedSubmodelElementCollection::ConnectedSubmodelElementCollection(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy):
  ConnectedSubmodelElement(proxy)
{}

void ConnectedSubmodelElementCollection::setValue(const basyx::objectCollection_t & value)
{
  this->setProxyValue(property::PropertyPaths::VALUE, value);
}

basyx::objectCollection_t ConnectedSubmodelElementCollection::getValue() const
{
  return this->getProxyCollection(property::PropertyPaths::VALUE);
}

void ConnectedSubmodelElementCollection::setOrdered(const bool & value)
{
  this->setProxyValue(SubmodelElementCollectionPaths::ORDERED, value);
}

bool ConnectedSubmodelElementCollection::isOrdered() const
{
  auto element = this->getProxy()->readElementValue(SubmodelElementCollectionPaths::ORDERED);
  return element.Get<bool>();
}

void ConnectedSubmodelElementCollection::setAllowDuplicates(const bool & value)
{
  this->setProxyValue(SubmodelElementCollectionPaths::ALLOWDUPLICATES, value);
}

bool ConnectedSubmodelElementCollection::isAllowDuplicates() const
{
  auto element = this->getProxy()->readElementValue(SubmodelElementCollectionPaths::ALLOWDUPLICATES);
  return element.Get<bool>();
}

void ConnectedSubmodelElementCollection::setElements(const basyx::objectMap_t & elements)
{
  this->setProxyValue(SubmodelPaths::SUBMODELELEMENT, elements);
}

basyx::objectMap_t ConnectedSubmodelElementCollection::getElements() const
{
  auto elements = this->getProxy()->readElementValue(SubmodelPaths::SUBMODELELEMENT);
  return elements.Get<basyx::objectMap_t>();
}


}
}
}
}